src/jit/decomposelongs.cpp

```diff
@@ -0,0 +1,1206 @@
+    NYI_IF(tree->gtOverflow(), "TYP_LONG cast with overflow");
+    switch (tree->AsCast()->CastFromType())
+    {
+    case TYP_INT:
+        if (tree->gtFlags & GTF_UNSIGNED)
+        {
+            loResult = tree->gtGetOp1();
+            hiResult = new (m_compiler, GT_CNS_INT) GenTreeIntCon(TYP_INT, 0);
+            m_compiler->fgSnipNode(curStmt, tree);
+        }
+        else
+        {
+            NYI("Lowering of signed cast TYP_INT->TYP_LONG");
+        }
+        break;
+
+    default:
+        NYI("Unimplemented type for Lowering of cast to TYP_LONG");
+        break;
+    }
+
+    FinalizeDecomposition(ppTree, data, loResult, hiResult);
+}
+
+
+//------------------------------------------------------------------------
+// DecomposeCnsLng: Decompose GT_CNS_LNG.
+//
+// Arguments:
+//    ppTree - the tree to decompose
+//    data - tree walk context
+//
+// Return Value:
+//    None.
+//
+void DecomposeLongs::DecomposeCnsLng(GenTree** ppTree, Compiler::fgWalkData* data)
+{
+    assert(ppTree != nullptr);
+    assert(*ppTree != nullptr);
+    assert(data != nullptr);
+    assert((*ppTree)->OperGet() == GT_CNS_LNG);
+
+    GenTree* tree = *ppTree;
+    INT32 hiVal = tree->AsLngCon()->HiVal();
+
+    GenTree* loResult = tree;
+    loResult->ChangeOperConst(GT_CNS_INT);
+    loResult->gtType = TYP_INT;
+
+    GenTree* hiResult = new (m_compiler, GT_CNS_INT) GenTreeIntCon(TYP_INT, hiVal);
+
+    FinalizeDecomposition(ppTree, data, loResult, hiResult);
+}
+
+
+//------------------------------------------------------------------------
+// DecomposeCall: Decompose GT_CALL.
+//
+// Arguments:
+//    ppTree - the tree to decompose
+//    data - tree walk context
+//
+// Return Value:
+//    None.
+//
+void DecomposeLongs::DecomposeCall(GenTree** ppTree, Compiler::fgWalkData* data)
+{
+    assert(ppTree != nullptr);
+    assert(*ppTree != nullptr);
+    assert(data != nullptr);
+    assert((*ppTree)->OperGet() == GT_CALL);
+
+    GenTree* parent = data->parent;
+
+    // We only need to force var = call() if the call is not a top-level node.
+    if (parent == nullptr)
+        return;
+
+    if (parent->gtOper == GT_STORE_LCL_VAR)
+    {
+        // If parent is already a STORE_LCL_VAR, we can skip it if
+        // it is already marked as lvIsMultiRegRet.
+        unsigned varNum = parent->AsLclVarCommon()->gtLclNum;
+        if (m_compiler->lvaTable[varNum].lvIsMultiRegRet)
+        {
+            return;
+        }
+        else if (!m_compiler->lvaTable[varNum].lvPromoted)
+        {
+            // If var wasn't promoted, we can just set lvIsMultiRegRet.
+            m_compiler->lvaTable[varNum].lvIsMultiRegRet = true;
+            return;
+        }
+    }
+
+    // Otherwise, we need to force var = call()
+    GenTree* tree = *ppTree;
+    GenTree** treePtr = nullptr;
+    parent = tree->gtGetParent(&treePtr);
+
+    assert(treePtr != nullptr);
+
+    GenTreeStmt* asgStmt = m_compiler->fgInsertEmbeddedFormTemp(treePtr);
```

Comment:
As far as I can tell this can result in the creation of a top level statement so yes, `CreateTemporary` is needed here. It's also possible that the use of `fgInsertEmbeddedFormTemp` here results in double decomposition and that needs to be handled in `CreateTemporary` as well.